### PR TITLE
Running Cypress in Docker

### LINKED ISSUE #1770: Run e2e tests against deployed environments
### Description
Run e2e tests against deployed environments. This should run against an environment that's deployed against the regular flow, and uses Okta Preview and the Experian demo account. We'll need to add `dev` back to the deployment pipeline for this to work.

### Proposed solution 
After deploying to training or demo, add a step to run the e2e tests, configured to point at the dev URL

### Additional context
This ticket was created in response to [this postmortem](https://github.com/usds/prime-simplereport-docs/pull/15)

We may need to add different profiles for running the e2e tests locally vs in deployed environments. Locally we'll want to use Wiremock instead of Okta preview, for instance.

This should be run before deploying a PR to production, rather than just doing it daily. We may need to add a flag enabling users to bypass e2e tests to respond to production incidents.

### LINKED ISSUE #3280: Write e2e workflow and the health-check action to utilize docker
Currently, we install and run cypress and other dependencies locally on the GitHub runners.

We want to dockerize our e2e test runs.

Doing this would reduce the number of ways we run e2e tests. Ideally, this would minimize effort on future maintenance and changes.

This ticket was spawned from a conversation with @nickclyde and @jdorothy.

### LINKED ISSUE #3637: Reconfigure `test` environment to be in the main deployment pipeline
`test` is being repurposed as our automation environment. The following tasks should be completed to facilitate this transition:

- Modify requisite GitHub Action files to add `test` to the automated deployment pipeline.
- Ensure `test` is deployed to with each push to Main
- Modify the "dibs" bot, if required, to remove `test` as a claimable entity.

### PR DESCRIPTION
**Pull request by alismx**
# Cypress PULL REQUEST

resolves #1770 
resolves #3280 
resolves #3637 

## Related Issue

- Clean up cypress implementation
- Make use of standard docker setup
- Increase integration testing coverage of a live environment

## Changes Proposed

- Run Cypress in docker
- Add Test to our ci pipeline
- Run Cypress against a remote environment
- Move Cypress directory out of the frontend directory
- Update Cypress tests to handle more variable data
- Handle cypress user having access to multiple facilities
- Update Okta mocks for local runs
- Update .env files for use in docker
- Parameterize okta request for authenication
- Block staging/prod deploys if cypress fails against test

## Additional Information

- Add swap space to the local run is pretty straightforward and seems to have had a positive impact on preventing Cypress from hanging and other failures, I used an open source repo to do this but could easily add a new action. Open to preferences and thoughts.
- It would be nice to decouple wiremock from running in the cypress container

## Testing

- Double check the deployStg file as it's hasn't officially been run with everything in place
- [Passing Cypress run against localhost.simplereport.gov](https://github.com/CDCgov/prime-simplereport/actions/runs/2222814403)
- [Passing Cypress run against test.simplereport.gov](https://github.com/CDCgov/prime-simplereport/actions/runs/2222653895)
- [Failing Cypress run against test.simplereport.gov](https://github.com/CDCgov/prime-simplereport/actions/runs/2222491028)
 
### Infrastructure
- [x] Consult the results of the `terraform-plan` job inside the "Terraform Checks" workflow run for this PR. Confirm that there are no unexpected changes!

### Security
- [x] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [x] Any dependencies introduced have been vetted and discussed

## Cloud
- [x] If there are changes that cannot be tested locally, this has been deployed to our Azure `test`, `dev`, or `pentest` environment for verification

----

### CODE REVIEW

**Comment by alismx** [comment]
Noticed this now has some conflicts, I'll get these resolved this afternoon.

**Comment by alismx** [comment]
Merge conflicts and a failing test resulting from that merge have been fixed!

**Review by emmastephenson — CHANGES_REQUESTED** [review]
Overall looks great! Excited to get this running on `test`.

Just a few requests around filenames and the readme.

File: .github/workflows/e2eLocal.yml
Comment (alismx) [reply .github/workflows/e2eLocal.yml:3]: Yes!

File: README.md
Comment (emmastephenson) [reply README.md:182]: Ah! This is what I get for not expanding the whole file :) If this is specifically in the e2e section I'm happy to keep it as-is.

Comment (alismx) [reply README.md:183]: Awesome, I updated the file names for those sample env files. I also added an additional note in the readme about the intended usage of the `.env.cypress.sample` and the `.env.cypress.remote.sample` files.

**Review by emmastephenson — APPROVED** [review]
LGTM!

File: .github/actions/health-checks/action.yml
Comment (rin-skylight) [inline .github/actions/health-checks/action.yml:35]: I might have missed this somewhere, but it seems like we purged the call to `make check` from the rest of what we're doing. Did I read that right? If so, was that intentional? The only other place that has it is the `rollback-application` action.

File: .github/workflows/e2eLocal.yml
Comment (rin-skylight) [reply .github/workflows/e2eLocal.yml:18]: I appreciate the tag. I would rather see us use our own action rather than pulling in an external one. Fewer external dependencies = fewer possible intrusion attempts for us to worry about. (That, and the repo we're citing doesn't seem to have an explicit license, which I don't want us to get in trouble for down the road.)

**Review by rin-skylight — CHANGES_REQUESTED** [review]
Thank you for all of your hard work on this. I would love to see a few team members test these changes locally on varying setups, just to make sure we don't inadvertently introduce something that breaks someone's workflow. That's always a good idea when dealing with docker-based changes. Aside from that, if we can take care of that external action dependency, I think we'll be good to go with this!

File: .github/actions/health-checks/action.yml
Comment (alismx) [reply .github/actions/health-checks/action.yml:35]: Great question! It's not required for the local run. 

I think it would be fine to add back into our cypress workflow that runs against test.simplereport.gov, but I don't think it's required for a couple of reasons. 

- Cypress runs it's own healthcheck against the environment that includes a check for the commit hash.
- The deployTest workflow that calls our cypress workflow runs similar checks called `wait-for-%-slot-release` and `wait-for-%-commit`.

File: .github/workflows/e2eLocal.yml
Comment (alismx) [reply .github/workflows/e2eLocal.yml]: I went ahead and created that action, going to resolve this.

File: .env.sample
Comment (nickclyde) [inline .env.sample:6]: What was the reasoning for removing this file? We do reference it [in the README](https://github.com/CDCgov/prime-simplereport#developing-locally), as these env vars are actually needed to run the app locally IIRC.

File: .github/workflows/e2eLocal.yml
Comment (nickclyde) [inline .github/workflows/e2eLocal.yml:24]: Whoa, does this ubuntu image have homebrew installed by default?

**Review by nickclyde — COMMENTED** [review]
Amazing work on this Alis!! Just a few questions below.

File: .github/workflows/e2eLocal.yml
Comment (alismx) [reply .github/workflows/e2eLocal.yml:24]: Yes! I was surprised too.

File: .env.sample
Comment (alismx) [reply .env.sample:6]: Ah, this was a mistake as I renamed it to the sample for cypress. I'll add this back!

**Review by nickclyde — DISMISSED** [review]

**Review by emmastephenson — DISMISSED** [review]
LGTM!

File: .github/actions/health-checks/action.yml
Comment (alismx) [reply .github/actions/health-checks/action.yml:35]: @rin-skylight Checking in on whether you think we have readiness check coverage for this or if you think I should add this back in.

**Comment by alismx** [comment]
If anybody wants to run cypress this locally, I'd be happy to hop on a call and walk through it.

File: .github/actions/health-checks/action.yml
Comment (rin-skylight) [reply .github/actions/health-checks/action.yml:35]: I would like to see this added back in, personally. I did another check of this, and this is called from our other `deployENV` actions...I missed that initially. I think leaving it in is necessary for the development environments. Otherwise, we may not be alerted to commits not being in the right place for dev testing.

Unless there's something I am missing, of course!

Comment (alismx) [reply .github/actions/health-checks/action.yml:35]: I'll add this back in today!

Comment (alismx) [reply .github/actions/health-checks/action.yml:35]: So, following back up on this, I'm only seeing the following jobs in the `deployENV` workflows and related actions.

- `make check-${{ env.DEPLOY_ENV }}-readiness`
- `make wait-for-${{ env.DEPLOY_ENV }}-slot-commit`
- `make wait-for-${{ env.DEPLOY_ENV }}-slot-readiness`

I'll go ahead and add it in because I don't see any downside but I'm not seeing this particular make command anywhere else.

Comment (alismx) [reply .github/actions/health-checks/action.yml:35]: I added this check into the cypress run!

**Comment by alismx** [comment]
I believe I've addressed all outstanding feedback. I appreciate all the thorough reviews and feedback. It should be ready for another look!

**Review by nickclyde — DISMISSED** [review]

**Review by emmastephenson — DISMISSED** [review]
LGTM! Though if you could coordinate with @rin-skylight on merging this after #3675 that would be great.

File: cypress/support/commands.js
Comment (nathancrtr) [inline cypress/support/commands.js:110]: Nice workaround! This should help the e2e tests behave more like they're idempotent, even though they're really not (when run locally, at least).

File: frontend/ci-git-fix.sh
Comment (georgehager) [inline frontend/ci-git-fix.sh:4]: What does this do? Is it for local run or docker?

Comment (alismx) [reply frontend/ci-git-fix.sh:4]: This is a workaround to a reported [bug](https://github.com/actions/runner/issues/2033), it resolves a "dubious" permissions error when running the frontend with docker. We run a git command on frontend startup and this is required to do that.

File: cypress/stubs/orgSignUp/mappings/api_v1_users-d2c41957-0331-4c20-a597-7002630980ab.json
Comment (georgehager) [inline cypress/stubs/orgSignUp/mappings/api_v1_users-d2c41957-0331-4c20-a597-7002630980ab.json:8]: this is a neat trick!

Comment (georgehager) [reply cypress/stubs/orgSignUp/mappings/api_v1_users-d2c41957-0331-4c20-a597-7002630980ab.json:8]: thought I'm realizing this is in an auto-generated file 🤣

File: cypress/README.md
Comment (georgehager) [inline cypress/README.md:102]: Super helpful!

File: cypress/stubs/orgSignUp/mappings/api_v1_users-d2c41957-0331-4c20-a597-7002630980ab.json
Comment (emmastephenson) [reply cypress/stubs/orgSignUp/mappings/api_v1_users-d2c41957-0331-4c20-a597-7002630980ab.json:8]: @georgehager mostly auto-generated! It's created by [WireMock](https://github.com/CDCgov/prime-simplereport/wiki/WireMock) but this `any-string` piece is hand-touched :)

Comment (alismx) [reply cypress/stubs/orgSignUp/mappings/api_v1_users-d2c41957-0331-4c20-a597-7002630980ab.json:8]: This was a manual update to allow for any email. Doing so lets us add more people and not need to clean the db every time.

File: .gitignore
Comment (georgehager) [inline .gitignore:17]: do we still need `frontend/cypress.env.json`?

Comment (alismx) [reply .gitignore:17]: Nope! I'll push an update.

Comment (alismx) [reply .gitignore]: Done.

File: cypress/stubs/orgSignUp/mappings/api_v1_users-d2c41957-0331-4c20-a597-7002630980ab.json
Comment (georgehager) [reply cypress/stubs/orgSignUp/mappings/api_v1_users-d2c41957-0331-4c20-a597-7002630980ab.json:8]: Neat! I saw in the test we type in `user.email` so that makes sense!

**Review by georgehager — APPROVED** [review]
Looks great! this will make running e2es locally so much easier 🙌

**Comment by georgehager** [comment]
@alismx did you decide not to add a command for running all e2e tests headless locally outside of docker? It's something I imagine wanting to do _occasionally_, so I don't think it's super critical and might not be worth the effort of adding another separate command since they can all be run in open mode.

**Comment by alismx** [comment]
> @alismx did you decide not to add a command for running all e2e tests headless locally outside of docker? It's something I imagine wanting to do _occasionally_, so I don't think it's super critical and might not be worth the effort of adding another separate command since they can all be run in open mode.

What do you say about making a follow-up ticket? I can write that up and link it here.

**Comment by georgehager** [comment]
> 

Yep I think that's great! Like I said, not super critical :grin:

File: cypress/support/commands.js
Comment (emyl3) [inline cypress/support/commands.js:36]: same question here as well

File: cypress/README.md
Comment (emyl3) [inline cypress/README.md:1]: thank you for so thoroughly updating this!

**Review by emyl3 — COMMENTED** [review]
Couple questions for you alis! Thank you for your all your work on this! 😭 🙌

File: .env.cypress.remote.sample
Comment (alismx) [reply .env.cypress.remote.sample:14]: Committing this value to the repo is fine. The client id is very easy to find in our auth flow as it gets passed around in the URL of the auth requests. We don't want to expose API keys or secrets.

File: cypress/README.md
Comment (alismx) [reply cypress/README.md:1]: :sweat_smile:

File: .env.cypress.remote.sample
Comment (emyl3) [reply .env.cypress.remote.sample:14]: Got it! Thank you @alismx :D

**Review by emyl3 — APPROVED** [review]
LGTM!! Thank you!! 😹

File: .github/workflows/deployDev.yml
Comment (emmastephenson) [inline .github/workflows/deployDev.yml:116]: Curious why we're no longer verifying the release for `dev`?

File: cypress/README.md
Comment (emmastephenson) [reply cypress/README.md:1]: +1, this is awesome! Once this is merged, would you mind also adding it to the wiki? (or just linking to this README somewhere on our getting started wiki page)

**Review by emmastephenson — COMMENTED** [review]
Looks awesome Alis, thanks for all your work on this! 🙌

File: .github/workflows/deployDev.yml
Comment (alismx) [reply .github/workflows/deployDev.yml:116]: This check is the first test of the cypress suite, which happens locally during the PR process and against the `test` environment once a merge goes through. This was only in place until we could run the whole suite against a remote environment in its entirety. 

I wouldn't mind this type of check for all of our remote environments, but it does take some time because of its dependence on spinning up cypress.

**Comment by alismx** [comment]
I can't seem to merge this until the sonar job is fixed.

File: .github/workflows/deployDev.yml
Comment (emmastephenson) [reply .github/workflows/deployDev.yml:116]: I see, thanks for the explanation!

**Comment by snesm** [comment]
@alismx, it looks like "test Expected — Waiting for status to be reported" is the hold up.  The sonar check is not "required"

**Comment by rin-skylight** [comment]
@snesm Sonar is the issue. There is a dependent check in the "test" workflow. PR is already in with the bypass.

**Comment by alismx** [comment]
@snesm @rin-skylight @emmastephenson Not sure what I'm missing, but it seems like this "test" check is only blocking this PR. I thought it was because of the Sonar issue failing, but now that we've merged in that sonar bypass, I believe that's been ruled out. Has anybody seen this before? Does anybody see any issues with the github workflows in this PR?

**Comment by rin-skylight** [comment]
@alismx This can happen on occasion. Creating a PR with a new branch, that has all of your existing changes, will likely be the easiest thing to kick the checks into gear. I've forced a re-run of the `test` workflow, but that reporting hook still might not function correctly.

**Pull request merged**
